proposals/infra/INF-NNNN-availability-diagnostics.md

```diff
@@ -0,0 +1,262 @@
+<!-- {% raw %} -->
+
+# Strict Availability Diagnostics
+
+* Proposal: [INF-NNNN](INF-NNNN-availability-diagnostics.md)
+* Author(s): [Chris Bieneman](https://github.com/llvm-beanz)
+* Sponsor: [Chris Bieneman](https://github.com/llvm-beanz)
+* Status: **Under Consideration**
+* Planned Version: 20YY (Clang-only)
+
+## Introduction
+
+Strict availability diagnostics introduces two new modes for diagnosing the use
+of unavailable shader APIs. Unavailable shader APIs are APIs that are exposed in
+HLSL code but are not available in the target shader stage or shader model
+version.
+
+## Motivation
+
+Today the enforcement of API availability is the responsibility of bytecode
+validators, runtimes and drivers operating on optimized shader code. Depending
+contextually on where this enforcement occurs this can result in errors being
+produced either late in compilation when mapping back to source locations is
+difficult or at runtime when the runtime or driver rejects the compiled shader.
+
+Strict availability diagnostics provides more actionable diagnostics for users
+by both having detailed error messages and accurate message locations. It also
+provides diagnostics early and allows implementations to generate meaningful
+diagnostics even without generation of debug information.
+
+> Note: The DXIL validator requires line table information to generate error
+> locations which can be inaccurate. Disabling line table information can have a
+> significant compile-time benefit.
+
+## Proposed solution

```

Comment:
Shall those proposals have a `Limits` sections?

For example, this wouldn't cover cases where the limitation comes from differences in the API we use:

For example:
```
groupshared int runtimeArray[0];
```

This is allowed in DXIL, unbounded, groupshared array which size is defined at runtime.
But this cannot be implemented in Vulkan/SPIR-V due to some restrictions that forbids runtime arrays for `groupshared` variables.
This not being a builtin/intrinsic, this attribute cannot be used, and we still need to have some ad-hoc availability diagnostic for such cases.

Reply:
Unless there is a common pattern these special cases will probably need to be handled separately by targeted diagnostics for each API that differs.

Reply:
This shouldn't be legal DXIL, but apparently we didn't prevent it in the compiler or validator.  I believe this is a bug.

> This is allowed in DXIL, unbounded, groupshared array which size is defined at runtime.

Have you seen this in the wild?  What do you mean when you say "size is defined at runtime"?  How is it defined at runtime?  I am not aware of any officially supported mechanism to do this.

Reply:
It's an extension of what I could find on the docs 😊
Context: https://github.com/microsoft/DirectXShaderCompiler/issues/3667#issuecomment-2025620965

But this is one specific example, just wanted to check if those specifications should have a "limits" of "issue" section. This way we can list tradeoffs or known limits to we still have to deal with.

Reply:
Yea. I think this is out of scope for this document. Resource types are kinda special in that we allow unsized arrays, but all resource handles need to be fully scalarized, so they aren't really generated as unbounded arrays.

I think for the purposes of this document we should keep the focus narrow to API availability not other broader capability differences.